### ISSUE TITLE
As a rule of thumb how many configuration shoud a task try out?

### ISSUE THREAD

**Issue by Motorrat** [automl/auto-sklearn#142]
As I have learned the metalearning feature has about 30 fixed configurations to try out. How many configuration should a autosklearn fit task try out as a rule of thumb? 
Some of my datasets let the loss function to come down to the minimum in the first 10-20 model runs, so most of the time after that is just wasted. Also I do not see much increase of the precision for the algorithms that run longer than 60 seconds. I am monitoring the progress via watching the logs `grep -h "Finished evaluating" atsklrn_tmp/*`
I am considering writing my code so that it starts with small values for time_left_for_this_task and per_run_time_limit and increasing them if the progress is not there. 

So actually I have two questions:
1. is it possible to tell early enough that per_run_time_limit is insufficient? 
2. what is the design philosophy in regard to how many runs should be completed before we can call it a reasonable effort?

**Comment by mfeurer** [comment]
Unfortunately, this is an open problem. My rule of thumb is to rather spent more computational power than too less and determine based on the CV score compared to a validation score whether auto-sklearn overfitted. On the other hand, the meta-learning configurations tried in the beginning can be quite good and actually represent the best configuration SMAC would find itself. To the best of my knowledge, there is no principled way to determine when optimization converged. The same holds for your first question. While we're quite successful at optimizing hyperparameters, the optimization algorithms introduce yet another level of parameters that have to be set correctly. A general heuristic you coud try is to run some basic scikit-learn models and set the timeout to be twice their runtime (to allow the preprocessing to finish). In contrast to the second question, there are groups working on this problem, but the solutions don't scale to auto-sklearn yet.

**Comment by Motorrat** [comment]
> A general heuristic you coud try is to run some basic scikit-learn models and set the timeout to be twice their runtime
This can be very handy, actually. Great suggestion! Where can I get the list of models that autosklearn has? You certainly have a list, even if as a list in the source file, haven't you?

**Comment by mfeurer** [comment]
Would this code snippet help?

    import autosklearn.pipeline.components.classification
    print autosklearn.pipeline.components.classification._classifiers

By this you get the actual component object, you can then run them in their default configuration.

**Comment by Motorrat** [comment]
Cool! this will do for per_run_time_limit.

Is there a programmatic way to get to the resulting loss function values from the various AutoSklearnClassifier seeds and their run number? What I can imagine to do is to stop the process after twice as many runs as it took to get to the lowest loss value. (or after 10x of runs or whatever). 

The background is that I would hate to let the process run for 24 hours, where it is actually done after 10 minutes. This is the case with many datasets I use. Yet I also have big ones that take as much as 24 hours to get to a decent value.

**Comment by Motorrat** [comment]
Here is the heuristic that gets >0.98 performance for the parallel_example.py

`    import autosklearn.pipeline.components.classification
    clfs=autosklearn.pipeline.components.classification._classifiers
    from time import time

    max_clf_time=1 # second
    for clf_name,clf_class in clfs.items():
        default = clf_class.get_hyperparameter_search_space().get_default_configuration()
        clf=clf_class(**default._values)
        t0 = time()
        clf.fit(X,y)
        classifier_time = time() - t0
        print(clf_name+" training time: "+str(classifier_time))
        if max_clf_time < classifier_time:
            max_clf_time = classifier_time

    queue_factor = 30 # least number of models to test in each seed
    half_cpu_cores = 4
    if queue_factor*half_cpu_cores < 100: # 100 models to test overall
        queue_factor=100/half_cpu_cores
    per_run_time_limit = int(3*max_clf_time) # times 3 for preprocessor and ensemble
    time_left_for_this_task = int(3*queue_factor*max_clf_time)

    processes = []
    for i in range(half_cpu_cores):
        p = multiprocessing.Process(target=spawn_classifier, args=(i,'digits',time_left_for_this_task,per_run_time_limit))
        p.start()
        processes.append(p)
    for p in processes:
        p.join()
`

**Comment by mfeurer** [comment]
I'm already working on a documented interface to access the outcome of function evaluations already performed by SMAC.

**Comment by Motorrat** [comment]
I have added the heuristic to determine the per_run_time_limit based on the actual run time for classifiers for a given dataset here https://github.com/Motorrat/auto-sklearn/blob/patch-1/example/redly.py

**Comment by mfeurer** [comment]
A brief section on this topic will available in the documentation of the next release of auto-sklearn. I leave this issue open as this is still a relevant, but unanswered question.

**Comment by tribeband** [comment]
I met with a bug if this is can be called like this, where the time_left_for_this_tast is set to 18000 secs (5 hours). Then this model is still running when at 10:46 pm while this model is started at 0:05:24 in the night.
![lADPDgQ9rF99M2jNC9DND8A_4032_3024 jpg_620x10000q90g](https://user-images.githubusercontent.com/47083168/66620983-11e1f480-ec15-11e9-968f-46b9a02b3a08.jpg)
![lADPDgQ9rF9_HmDNC9DND8A_4032_3024 jpg_620x10000q90g](https://user-images.githubusercontent.com/47083168/66620991-17d7d580-ec15-11e9-8e8c-4371e832af04.jpg)
![lADPDgQ9rF99M17NC9DND8A_4032_3024 jpg_620x10000q90g](https://user-images.githubusercontent.com/47083168/66620994-1c9c8980-ec15-11e9-9b5c-05ee6c9a8f2d.jpg)

**Comment by tribeband** [comment]
Finally the auto-sklearn terminated at 14:00 getting a 0.5 AUC_ROC score for this unbalanced binary-classification example where the label ratio is 96:4.
